CSS timing functions should not eagerly evaluate calc()

weinig opened the pull request:
#### 2047aa6f9356073f2eac55f0ea91d31bd024717a
<pre>
CSS timing functions should not eagerly evaluate calc()
<a href="https://bugs.webkit.org/show_bug.cgi?id=278925">https://bugs.webkit.org/show_bug.cgi?id=278925</a>

Reviewed by Darin Adler.

The CSS timing functions were using the RawResolver to eagerly evaluate calc()
during parsing, which results in incorrect serialization of specified values and
doesn&apos;t allow for use of relative lengths in the calculations. Instead, calc()
evaluation should be delayed until style building.

This change updates the timing function CSS values to store unevaluated primitives
and resolve calc() at style building time. It still doesn&apos;t allow for use of
relative lengths, as not all call sites can trivially pass in conversion data, so
that is deferred to a follow up.

While making this change, parsing has been extracted into its own file and
resolution to TimingFunction types has been moved out of platform/ (which was a
layering violation) and into CSSTimingFunctionValues.h/cpp. Additionally, a
helper to parse timing functions from strings has been added so that callers
don&apos;t have to invoke the entire CSS parser when only timing functions are needed.

* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/linear-timing-functions-syntax-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/linear-timing-functions-syntax.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-valid.html:
    - Update timing function tests with cases that involve calc().

* Source/WTF/wtf/text/TextStream.cpp:
(WTF::TextStream::operator&lt;&lt;):
* Source/WTF/wtf/text/TextStream.h:
    - Add support for FormattedCSSNumber to TextStream. Remove
      unnecessary allocation when serializing HexNumberBuffer.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add files: CSSPropertyParserConsumer+TimingFunction.h/cpp

* Source/WebCore/animation/AnimationEffect.cpp:
* Source/WebCore/animation/AnimationEffect.h:
* Source/WebCore/animation/AnimationEffect.idl:
* Source/WebCore/animation/CustomEffect.cpp:
* Source/WebCore/animation/CustomEffect.h:
* Source/WebCore/animation/CustomEffect.idl:
* Source/WebCore/animation/DocumentTimeline.cpp:
* Source/WebCore/animation/KeyframeEffect.cpp:
    - Pipe the Document around so that callers switching from
      TimingFunction::createFromCSSText to CSSPropertyParserHelpers::parseTimingFunction
      can have a valid CSSParserContext. This is important since it dictates if spring()
      is allowed or not.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
    - Add helper, isOne(), to test if the value is exactly 1, returns
      std::nullopt for calc() like similar functions.

* Source/WebCore/css/CSSTimingFunctionValue.cpp:
(WebCore::createTimingFunction):
    - Move logic from TimingFunction::createFromCSSText to CSSTimingFunctionValue.
      This centralizes things and avoids the layering violation.

(WebCore::CSSLinearTimingFunctionValue::createTimingFunction const):
    - Rather than running the &quot;create-a-linear-easing-function&quot; at parse time,
      which is not possible in the presence of calc(), it is now deferred to
      style building time and moved here.

* Source/WebCore/css/CSSTimingFunctionValue.h:
    - Store the parsed forms of the parameters to allow late resolution of
      calc at style building.

* Source/WebCore/css/CSSToStyleMap.cpp:
* Source/WebCore/css/CSSToStyleMap.h:
    - Forward to the new Style::BuilderConverter::convertTimingFunction().

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForAnimationTimingFunction):
    - Update computed value creation to create the appropriate primitive
      value parameters for the easing functions.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
    - Add include for call to consumeTimingFunction.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimingFunction.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimingFunction.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
    - Moved consumer implementations from CSSPropertyParserHelpers, and rework
      them by removing the eager evaluation of calc via RawResolver.

* Source/WebCore/css/process-css-properties.py:
    - Add include for generated callers of consumeTimingFunction().

* Source/WebCore/platform/animation/TimingFunction.cpp:
* Source/WebCore/platform/animation/TimingFunction.h:
(WebCore::TimingFunction::cssText const):
(WebCore::TimingFunction::createFromCSSText): Deleted.
(WebCore::TimingFunction::createFromCSSValue): Deleted.
    - Moved code directly dealing with CSS types to CSSTimingFunctionValue.h/cpp.
    - The serialization code remains and has been update to match CSS semantics by
      using FormattedCSSNumber for numerics. Ideally this would go away as well
      with callers using the serialization in CSSTimingFunctionValue, but that is
      an existing issue and requires a bit of refactoring that was too much for this
      change as is.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTimingFunction):
    - Add converter for TimingFunction which forwards to createTimingFunction().

* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::keyframeRulesForName const):
(WebCore::Style::Resolver::keyframeStylesForAnimation):
    - Replace use of TimingFunction::createFromCSSValue() with its replacement
      createTimingFunction().

Canonical link: <a href="https://commits.webkit.org/283036@main">https://commits.webkit.org/283036@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d13db8d19fc4e9433fb99c777d6957c6b566ccd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64984 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44351 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17598 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69008 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15590 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67102 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52134 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15872 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52206 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10763 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68050 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41025 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56242 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32828 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37679 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13613 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14466 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/58097 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59591 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13951 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70713 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/64227 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8936 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13428 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59535 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8968 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56302 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59757 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14343 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7371 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1047 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/85993 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40163 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15157 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41240 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42421 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40984 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/TextStream.cpp:
```diff
@@ -145,7 +145,13 @@ TextStream& TextStream::operator<<(StringView string)
 
 TextStream& TextStream::operator<<(const HexNumberBuffer& buffer)
 {
-    m_text.append(makeString(buffer));
```
Good catch. Annoying we had left it like this.

weinig committed:
CSS timing functions should not eagerly evaluate calc()
https://bugs.webkit.org/show_bug.cgi?id=278925

Reviewed by Darin Adler.

The CSS timing functions were using the RawResolver to eagerly evaluate calc()
during parsing, which results in incorrect serialization of specified values and
doesn't allow for use of relative lengths in the calculations. Instead, calc()
evaluation should be delayed until style building.

This change updates the timing function CSS values to store unevaluated primitives
and resolve calc() at style building time. It still doesn't allow for use of
relative lengths, as not all call sites can trivially pass in conversion data, so
that is deferred to a follow up.

While making this change, parsing has been extracted into its own file and
resolution to TimingFunction types has been moved out of platform/ (which was a
layering violation) and into CSSTimingFunctionValues.h/cpp. Additionally, a
helper to parse timing functions from strings has been added so that callers
don't have to invoke the entire CSS parser when only timing functions are needed.

* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/linear-timing-functions-syntax-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/linear-timing-functions-syntax.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-easing/timing-functions-syntax-valid.html:
    - Update timing function tests with cases that involve calc().

* Source/WTF/wtf/text/TextStream.cpp:
(WTF::TextStream::operator<<):
* Source/WTF/wtf/text/TextStream.h:
    - Add support for FormattedCSSNumber to TextStream. Remove
      unnecessary allocation when serializing HexNumberBuffer.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add files: CSSPropertyParserConsumer+TimingFunction.h/cpp

* Source/WebCore/animation/AnimationEffect.cpp:
* Source/WebCore/animation/AnimationEffect.h:
* Source/WebCore/animation/AnimationEffect.idl:
* Source/WebCore/animation/CustomEffect.cpp:
* Source/WebCore/animation/CustomEffect.h:
* Source/WebCore/animation/CustomEffect.idl:
* Source/WebCore/animation/DocumentTimeline.cpp:
* Source/WebCore/animation/KeyframeEffect.cpp:
    - Pipe the Document around so that callers switching from
      TimingFunction::createFromCSSText to CSSPropertyParserHelpers::parseTimingFunction
      can have a valid CSSParserContext. This is important since it dictates if spring()
      is allowed or not.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
    - Add helper, isOne(), to test if the value is exactly 1, returns
      std::nullopt for calc() like similar functions.

* Source/WebCore/css/CSSTimingFunctionValue.cpp:
(WebCore::createTimingFunction):
    - Move logic from TimingFunction::createFromCSSText to CSSTimingFunctionValue.
      This centralizes things and avoids the layering violation.

(WebCore::CSSLinearTimingFunctionValue::createTimingFunction const):
    - Rather than running the "create-a-linear-easing-function" at parse time,
      which is not possible in the presence of calc(), it is now deferred to
      style building time and moved here.

* Source/WebCore/css/CSSTimingFunctionValue.h:
    - Store the parsed forms of the parameters to allow late resolution of
      calc at style building.

* Source/WebCore/css/CSSToStyleMap.cpp:
* Source/WebCore/css/CSSToStyleMap.h:
    - Forward to the new Style::BuilderConverter::convertTimingFunction().

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForAnimationTimingFunction):
    - Update computed value creation to create the appropriate primitive
      value parameters for the easing functions.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
    - Add include for call to consumeTimingFunction.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimingFunction.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimingFunction.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
    - Moved consumer implementations from CSSPropertyParserHelpers, and rework
      them by removing the eager evaluation of calc via RawResolver.

* Source/WebCore/css/process-css-properties.py:
    - Add include for generated callers of consumeTimingFunction().

* Source/WebCore/platform/animation/TimingFunction.cpp:
* Source/WebCore/platform/animation/TimingFunction.h:
(WebCore::TimingFunction::cssText const):
(WebCore::TimingFunction::createFromCSSText): Deleted.
(WebCore::TimingFunction::createFromCSSValue): Deleted.
    - Moved code directly dealing with CSS types to CSSTimingFunctionValue.h/cpp.
    - The serialization code remains and has been update to match CSS semantics by
      using FormattedCSSNumber for numerics. Ideally this would go away as well
      with callers using the serialization in CSSTimingFunctionValue, but that is
      an existing issue and requires a bit of refactoring that was too much for this
      change as is.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTimingFunction):
    - Add converter for TimingFunction which forwards to createTimingFunction().

* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::keyframeRulesForName const):
(WebCore::Style::Resolver::keyframeStylesForAnimation):
    - Replace use of TimingFunction::createFromCSSValue() with its replacement
      createTimingFunction().

Canonical link: https://commits.webkit.org/283036@main
